File: leafy/Services/Cloudflare/CloudflareTimetableSharingService.swift
import Foundation
import OSLog

nonisolated protocol TimetableSharing: Sendable {
    func publishSnapshot(courses: [SharedTimetableCourse]) async throws -> SharedTimetableSnapshot
    @discardableResult func publishExistingSnapshotIfNeeded(courses: [SharedTimetableCourse]) async -> Bool
    func fetchMySnapshot() async throws -> SharedTimetableSnapshot?
    func fetchViewableSnapshots() async throws -> [SharedTimetableSnapshot]
    func fetchMyShareMembers() async throws -> [TimetableShareMember]
    func fetchMyInvites() async throws -> [TimetableInvite]
    func createInvite() async throws -> TimetableInvite
    func acceptInvite(code: String) async throws -> SharedTimetableSnapshot
    func revokeShare(viewerID: UUID) async throws
    func stopSharing() async throws
    func leaveShare(ownerID: UUID) async throws
}

nonisolated struct CloudflareTimetableSharingService: TimetableSharing {
    private let backend = CloudflareCommunityRepository()
    private func requireProfile() async throws -> CommunityProfile {
        try await backend.ensureAnonymousSession()
        guard let profile = try await backend.fetchCurrentProfile() else { throw TimetableSharingError.missingProfile }
        guard profile.isProfileComplete else { throw TimetableSharingError.profileCompletionRequired }
        return profile
    }
    func publishSnapshot(courses: [SharedTimetableCourse]) async throws -> SharedTimetableSnapshot {
        guard !courses.isEmpty else { throw TimetableSharingError.emptySnapshot }
        _ = try await requireProfile()
        let semester = await SemesterConfig.refreshRemoteIfAvailable()
        struct Input: Encodable { let semester_id: String; let courses: [SharedTimetableCourse] }
        return try await backend.request("/v1/timetables", method: "PUT", body: Input(semester_id: semester.semesterID, courses: courses))
    }
    @discardableResult
    func publishExistingSnapshotIfNeeded(courses: [SharedTimetableCourse]) async -> Bool {
        guard !courses.isEmpty else { return false }
        do {
            guard try await fetchMySnapshot() != nil else { return false }
            _ = try await publishSnapshot(courses: courses)
            return true
        } catch {
            CommunityDiagnostics.log.error("Cloudflare shared timetable refresh failed: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }
    func snapshots() async throws -> [SharedTimetableSnapshot] {
        let semester = await SemesterConfig.refreshRemoteIfAvailable()
        return try await backend.get("/v1/timetables", backend.query(["semester_id":semester.semesterID]))
    }
    func fetchMySnapshot() async throws -> SharedTimetableSnapshot? {
        let profile = try await requireProfile()
        return try await snapshots().first { $0.ownerID == profile.id }
    }
    func fetchViewableSnapshots() async throws -> [SharedTimetableSnapshot] {
        let profile = try await requireProfile()
        return try await snapshots().filter { $0.ownerID != profile.id }
    }
    func fetchMyShareMembers() async throws -> [TimetableShareMember] {
        _ = try await requireProfile()
        return try await backend.get("/v1/timetables/members")
    }
    func fetchMyInvites() async throws -> [TimetableInvite] {
        _ = try await requireProfile()
        return try await backend.get("/v1/timetables/invites")
    }
    func createInvite() async throws -> TimetableInvite {
        _ = try await requireProfile()
        return try await backend.request("/v1/timetables/invites", body: CloudflareCommunityRepository.Empty())
    }
    func acceptInvite(code: String) async throws -> SharedTimetableSnapshot {
        _ = try await requireProfile()
        return try await backend.request("/v1/timetables/invites/accept", body: ["code":TimetableSharingService.normalizeInviteCode(code)])
    }
    func revokeShare(viewerID: UUID) async throws {
        let members = try await fetchMyShareMembers()
        guard let member = members.first(where: { $0.viewerID == viewerID }) else { throw TimetableSharingError.notShareOwner }
        try await backend.perform("/v1/timetables/members/\(member.id)", method: "DELETE")
    }
    func stopSharing() async throws { try await backend.perform("/v1/timetables/stop") }
    func leaveShare(ownerID: UUID) async throws {
        let members: [TimetableShareMember] = try await backend.get("/v1/timetables/members", backend.query(["direction":"viewing"]))
        guard let member = members.first(where: { $0.ownerID == ownerID }) else { throw TimetableSharingError.notShareOwner }
        try await backend.perform("/v1/timetables/members/\(member.id)/leave")
    }
}

File: android/app/src/main/java/com/myleafy/android/services/cloudflare/CloudflareTimetableSharingService.kt
package com.myleafy.android.services.cloudflare

import com.myleafy.android.services.TimetableSharingService
import com.myleafy.android.services.supabase.*
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.put

class CloudflareTimetableSharingService(private val client: MyLeafyBackendClient) : TimetableSharingService {
    override suspend fun isBackendAvailable(): Boolean {
        client.request("/v1/timetables")
        return true
    }

    override suspend fun publish(campusId: String, ownerId: String, semesterId: String, courses: List<SharedTimetableCourseDto>): SharedTimetableSnapshotDto {
        require(courses.isNotEmpty()) { "本地课表为空，请先同步课表" }
        return client.decodeRequest("/v1/timetables", "PUT", buildJsonObject {
            put("semester_id", semesterId); put("courses", backendJson.encodeToJsonElement(courses))
        })
    }

    private suspend fun snapshots(semesterId: String): List<SharedTimetableSnapshotDto> =
        client.decodeRequest("/v1/timetables", query = mapOf("semester_id" to semesterId))

    override suspend fun mySnapshot(campusId: String, ownerId: String, semesterId: String): SharedTimetableSnapshotDto? =
        snapshots(semesterId).firstOrNull { it.owner_id == ownerId }

    override suspend fun viewableSnapshots(campusId: String, ownerId: String, semesterId: String): List<SharedTimetableSnapshotDto> =
        snapshots(semesterId).filter { it.owner_id != ownerId }

    override suspend fun members(campusId: String, ownerId: String): List<TimetableShareMemberDto> =
        client.decodeRequest("/v1/timetables/members")

    override suspend fun invites(campusId: String, ownerId: String): List<TimetableInviteDto> =
        client.decodeRequest("/v1/timetables/invites")

    override suspend fun createInvite(): TimetableInviteDto = client.decodeRequest("/v1/timetables/invites", "POST")
    override suspend fun accept(code: String): SharedTimetableSnapshotDto =
        client.decodeRequest("/v1/timetables/invites/accept", "POST", buildJsonObject { put("code", TimetableSharingService.normalizeCode(code)) })

    override suspend fun revoke(ownerId: String, viewerId: String) {
        val memberships = client.decodeRequest<List<TimetableShareMemberDto>>("/v1/timetables/members")
        val member = memberships.firstOrNull { it.owner_id == ownerId && it.viewer_id == viewerId }
            ?: error("共享关系不存在或已撤销")
        client.request("/v1/timetables/members/${apiID(member.id)}", "DELETE")
    }

    override suspend fun stopSharing() { client.request("/v1/timetables/stop", "POST") }
    override suspend fun leave(ownerId: String) {
        val memberships = client.decodeRequest<List<TimetableShareMemberDto>>("/v1/timetables/members", query = mapOf("direction" to "incoming"))
        val member = memberships.firstOrNull { it.owner_id == ownerId } ?: error("共享关系不存在或已撤销")
        client.request("/v1/timetables/members/${apiID(member.id)}/leave", "POST")
    }
}
